streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java

```diff
@@ -0,0 +1,366 @@
+                    throw new StreamsException("Sent one request but received multiple or no responses.");
+                }
+                if (responseList.get(0).request().equals(clientRequest)) {
+                    return responseList.get(0);
+                } else {
+                    throw new StreamsException("Inconsistent response received.");
+                }
+            }
+        }
+        throw new StreamsException("Failed to get response from broker within timeout");
+    }
+
+
+    /**
+     * Get the metadata for a topic.
+     * @param topic
+     * @return
+     */
+    public MetadataResponse.TopicMetadata getTopicMetadata(final String topic) {
+
+        final ClientResponse clientResponse = sendRequest(MetadataRequest.allTopics().toStruct(), ApiKeys.METADATA);
+        final MetadataResponse metadataResponse = new MetadataResponse(clientResponse.responseBody());
+        for (MetadataResponse.TopicMetadata topicMetadata: metadataResponse.topicMetadata()) {
+            if (topicMetadata.topic().equalsIgnoreCase(topic)) {
+                return topicMetadata;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Check to see if a topic exists.
+     * @param topicName
+     * @return
+     */
+    public boolean topicExists(final String topicName) {
+
+        if (getTopicMetadata(topicName) != null) {
+            return true;
+        }
+        return false;
+    }
+
+    public Set<InternalTopicConfig> filterValidateExistingTopics(final Map<InternalTopicConfig, Integer> topicsPartitionsMap) {
+        Set<InternalTopicConfig> nonExistingTopics = new HashSet<>();
+        final ClientResponse clientResponse = sendRequest(MetadataRequest.allTopics().toStruct(), ApiKeys.METADATA);
+        final MetadataResponse metadataResponse = new MetadataResponse(clientResponse.responseBody());
+        // The names of existing topics
+        Map<String, Integer> existingTopicNamesPartitions = new HashMap<>();
+        for (MetadataResponse.TopicMetadata topicMetadata: metadataResponse.topicMetadata()) {
+            existingTopicNamesPartitions.put(topicMetadata.topic(), topicMetadata.partitionMetadata().size());
+        }
+        // Add the topics that don't exist to the nonExistingTopics.
+        for (InternalTopicConfig topic: topicsPartitionsMap.keySet()) {
+            if (existingTopicNamesPartitions.get(topic.name()) == null) {
+                nonExistingTopics.add(topic);
+            } else if (existingTopicNamesPartitions.get(topic.name()) != topicsPartitionsMap.get(topic.name())) {
+
+            }
+        }
+
+        return nonExistingTopics;
+    }
+
+    /**
+     * Return the non existing topics.
+     *
+     * @param topicsPartitionsMap
+     * @param topicsMetadata
+     * @return
+     */
+    public Map<InternalTopicConfig, Integer> filterExistingTopics(final Map<InternalTopicConfig, Integer> topicsPartitionsMap, Collection<MetadataResponse.TopicMetadata> topicsMetadata) {
```

Comment:
Can we merge these two functions `filterExistingTopics` and `getTopicsToBeDeleted` into a single one, or just in-line this logic in the `makeReady` function as this seems specific in the InternalTopicManager, not in the KafkaClient.